sensor/common/sensor/sensor.go

```diff
@@ -411,6 +416,7 @@ func (s *Sensor) notifySyncDone(syncDone *concurrency.Signal, centralCommunicati
 	case <-syncDone.Done():
 		s.currentStateMtx.Lock()
 		defer s.currentStateMtx.Unlock()
+		s.changeStateNoLock(common.SensorComponentEventCentralReachable)
 		s.notifyAllComponents(common.SensorComponentEventSyncFinished)
```

Comment:
Well, that is an idea to consider... However, I am not fully sure yet whether a merge would be the correct thing to do (but also not saying that it is incorrect). Here is my train of thoughts:

1. Ping to Central is a false-friend, as it is done over HTTP. We can use it as a necessary condition for the central communication to be up, but it is not a sufficient condition - I saw few cases when ping was okay, but gRPC had problems.
2. SyncFinished is a valuable information - we know that there is a working connection to Central AND an important operation of sync has finished. Sync is an indication that connection is okay, but a signal that connection is okay is not an indication that sync finished.
3. What I actually meant with this change was to mark Central as Reachable as soon as the first communication over gRPC connection finishes with success. SyncFinished was a comfortable thing to pick, but actually a much better moment would be when the Hello protocol is finished with success.

So, I would rather suggest to keep the Sync Finished as it is and maybe add another event `HelloComplete` and go online immediately after it? We could also name the `HelloComplete` as `CentralReachable`, although we may think of better name to avoid confusion with HTTP-Central-Ping.

The idea is great and we should do this, but I would like to avoid merging states when we may loose some information about the Sensor state. Does that make sense? :)

Reply:
I think it all boils down to what the components need to know, or in other words: 

- Does any component need to know when Central is reachable (as in ping succeeded)? 
- Does any component need to know when the hello message is successfully sent?

If the answer is no to both, then adding a new state for `HelloComplete` is not really worth it IMO and I'd remove the `CentralReachable` status. Just keep `SyncFinished` and switch the components to online then.

If the answer is yes to the first question, then I'd keep the `CentralReachable` where it is right now (right after the ping) and I'd change the components that do not need to react to `CentralReachable` to react to `SyncFinished` instead.

If the answer is yes to the second question, then I'd add (like you suggested) a new state `HelloComplete` and make the necessary components react to it.

Alternatively if we see Sensor as a state machine with components that plug into it, we could provide the three states regardless of whether the components use them.

Reply:
Your analysis makes sense, but to provide answers to those two questions, we would need to go over the list of components and check the details for each of them. We may do this, but I am not sure this should be a prerequisite for this fix.

The problems is that `CentralReachable` checks if Central is Reachable over HTTP, whereas the components rely on Central being reachable over gRPC. There is a possibility that Central is reachable over HTTP but not over gRPC. So my proposal boils down to introduction of `CentralReachableGRPC` that would be emitted shortly after the Hello protocol completes. The Hello protocol is not strictly required here - Central maybe gRPC reachable earlier but we would need to write some more code to find that moment - but it is convenient, as  we know that Hello runs over gRPC and when Hello completes, then gRPC must be up and running.

Reply:
So for now, the situation looks like this:

1. Central responds to HTTP ping - `CentralReachable`
2. Central finishes Hello protocol - no signal
3. Sync starts - no signal
4. Sync finishes - `SyncFinished`

This is how we could make it correct:

1. Central responds to HTTP ping - `CentralReachableHTTP`
2. Central finishes Hello protocol - `CentralReachableGRPC`
3. Sync starts - no signal
4. Sync finishes - `SyncFinished`

but for convenience, we may use `SyncFinished` as `CentralReachableGRPC` and save us some code.
If we find a scenario where some components could benefit from starting sooner than `SyncFinished`, we may add the `CentralReachableGRPC`.

For now, I will rely on `SyncFinished` for simplicity. 
However, I would like rename the `CentralReachable` to `CentralReachableHTTP` to avoid confusion.

Reply:
> 1. Central responds to HTTP ping - CentralReachableHTTP
> 2. Central finishes Hello protocol - CentralReachableGRPC
> 3. Sync starts - no signal
> 4. Sync finishes - SyncFinished

I agree with this, the only thing I'm not sure is the usability ATM of `CentralReachableHTTP/GRPC`. I don't think that currently we have components that would need those signals, but it's true that we would have to look at all of them to really know for sure and, it is also true that, we might have components that might need those signals in the future.